src/libmongoc/tests/test-mongoc-collection.c

```diff
@@ -6350,6 +6350,76 @@ test_remove_multi (void)
    mongoc_client_destroy (client);
 }
 
+static void
+test_timeout_ms (void)
+{
+   mongoc_client_t *client = mongoc_client_new ("mongodb://host/db_name");
+   mongoc_database_t *db = NULL;
+   mongoc_collection_t *coll =
+      mongoc_client_get_collection (client, "db", "test");
+   mongoc_collection_t *coll2 = NULL;
+   bool res;
+   bson_error_t error;
+
+   /* TODO CDRIVER-3850 no timeout MS returns client's timeout */
+   /* BSON_ASSERT (mongoc_client_set_timeout_ms (client, 10, &error));
+      ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll), ==,
+      mongoc_client_get_timeout_ms (client)); */
+
+   BSON_ASSERT (mongoc_collection_get_timeout_ms (coll) ==
+                MONGOC_TIMEOUTMS_UNSET);
+
+   /* negative timeouts are invalid */
+   res = mongoc_collection_set_timeout_ms (coll, -1, &error);
+   BSON_ASSERT (!res);
+   ASSERT_ERROR_CONTAINS (error,
+                          MONGOC_ERROR_TIMEOUT,
+                          MONGOC_ERROR_TIMEOUT_INVALID,
+                          "timeoutMS must be a non-negative integer");
+
+   /* successful timeouts */
+   ASSERT_OR_PRINT (mongoc_collection_set_timeout_ms (coll, 300, &error),
+                    error);
+   ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll), ==, 300);
+
+   ASSERT_OR_PRINT (mongoc_collection_set_timeout_ms (coll, 100, &error),
+                    error);
+   ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll), ==, 100);
+
+   ASSERT_OR_PRINT (mongoc_collection_set_timeout_ms (coll, 0, &error), error);
+   ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll), ==, 0);
+
+   BSON_ASSERT (mongoc_collection_get_timeout_ms (coll) !=
+                MONGOC_TIMEOUTMS_UNSET);
+
+   /* Integer limits */
+   ASSERT_OR_PRINT (mongoc_collection_set_timeout_ms (coll, INT64_MAX, &error),
+                    error);
+   ASSERT_CMPINT64 (mongoc_collection_get_timeout_ms (coll), ==, INT64_MAX);
+   BSON_ASSERT (!mongoc_collection_set_timeout_ms (coll, INT64_MIN, &error));
+
+   /* When a database is involved, inheritance chain changes */
+   db = mongoc_client_get_database (client, "test");
+   ASSERT_OR_PRINT (mongoc_database_set_timeout_ms (db, 50, &error), error);
+
+   coll2 = mongoc_database_get_collection (db, "test");
+   ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll2),
+                  ==,
+                  mongoc_database_get_timeout_ms (db));
+
+   /* We do not track parent db timeout changes over time, though */
+   ASSERT_OR_PRINT (mongoc_database_set_timeout_ms (db, 200, &error), error);
+   ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll2), ==, 50);
+
+   ASSERT_OR_PRINT (mongoc_collection_set_timeout_ms (coll2, 300, &error),
+                    error);
+   ASSERT_CMPINT (mongoc_collection_get_timeout_ms (coll2), ==, 300);
```

Comment:
Duplicate tests? This same thing is done on lines 6381-6383

Reply:
These lines are to test that the collection doesn't keep using the timeout set on its parent `db` on L6411.  It's just a coincidence that it uses the same value `300`, it does not matter what value it uses as long as it's different from the parent `db`'s timeoutMS.